stage0/src/dice.rs

```diff
@@ -4,30 +4,136 @@
 
 use crate::image_header::Image;
 use crate::Handoff;
+use core::mem;
 use dice_crate::{
-    AliasCertBuilder, AliasData, AliasOkm, Cdi, CdiL1, CertSerialNumber,
-    DeviceIdOkm, RngData, RngSeed, SeedBuf, SerialNumber, SpMeasureCertBuilder,
-    SpMeasureData, SpMeasureOkm, TrustQuorumDheCertBuilder, TrustQuorumDheOkm,
+    AliasCertBuilder, AliasData, AliasOkm, Cdi, CdiL1, Cert, CertData,
+    CertSerialNumber, DeviceIdCertBuilder, DeviceIdOkm, RngData, RngSeed,
+    SeedBuf, SerialNumber, SizedBlob, SpMeasureCertBuilder, SpMeasureData,
+    SpMeasureOkm, TrustQuorumDheCertBuilder, TrustQuorumDheOkm,
 };
 use lpc55_pac::Peripherals;
-use salty::signature::Keypair;
+use lpc55_puf::Puf;
+use salty::{constants::SECRETKEY_SEED_LENGTH, signature::Keypair};
 
-#[cfg(feature = "dice-self")]
-use crate::dice_mfg_self::gen_mfg_artifacts;
 #[cfg(feature = "dice-mfg")]
-use crate::dice_mfg_usart::gen_mfg_artifacts;
+use crate::dice_mfg_usart;
 
-fn gen_deviceid_keypair(cdi: &Cdi) -> Keypair {
-    let devid_okm = DeviceIdOkm::from_cdi(cdi);
+pub const SEED_LEN: usize = SECRETKEY_SEED_LENGTH;
+pub const KEYCODE_LEN: usize =
+    Puf::key_to_keycode_len(SEED_LEN) / mem::size_of::<u32>();
+pub const KEY_INDEX: u32 = 1;
 
-    Keypair::from(devid_okm.as_bytes())
-}
-
-pub struct SerialNumbers {
+/// Data we get back from the manufacturing process.
+pub struct MfgResult {
     pub cert_serial_number: CertSerialNumber,
     pub serial_number: SerialNumber,
+    pub persistid_keypair: Keypair,
+    pub persistid_cert: SizedBlob,
+    pub intermediate_cert: SizedBlob,
 }
 
+/// Generate stuff associated with the manufacturing process.
+fn gen_mfg_artifacts(peripherals: &Peripherals) -> MfgResult {
+    // If configured to go through the manufacturing process (feature
+    // "dice-mfg" we call the appropriate function from the
+    cfg_if::cfg_if! {
+        if #[cfg(feature = "dice-mfg")] {
+            dice_mfg_usart::gen_mfg_artifacts_usart(peripherals)
+        } else if #[cfg(feature = "dice-self")] {
+            gen_mfg_artifacts_self(peripherals)
+        }
```

Comment:
Do we always expect one or the other?

If so, putting an `else { compile_error!(...) ` would be helpful.

Reply:
In the scope of this file: yes. It's valid to have both disabled, however that's handled in the main module. I'll make this fix as well as cleaning up the fragment of a comment above it :facepalm: